certbot/certbot/_internal/renewal.py

```diff
@@ -290,21 +303,66 @@ def _restore_str(name, value):
     return None if value == "None" else value
 
 
-def should_renew(config, lineage):
+def should_renew(config: configuration.NamespaceConfig, lineage: interfaces.RenewableCert) -> bool:
     """Return true if any of the circumstances for automatic renewal apply."""
     if config.renew_by_default:
         logger.debug("Auto-renewal forced with --force-renewal...")
         return True
     if lineage.should_autorenew():
         logger.info("Certificate is due for renewal, auto-renewing...")
         return True
+    if util.is_staging(config.server) and _ari_should_renew(config, lineage):
+        logger.info("Suggested renewal time in the past, renewing...")
+        return True
     if config.dry_run:
         logger.info("Certificate not due for renewal, but simulating renewal for dry run")
         return True
     display_util.notify("Certificate not yet due for renewal")
     return False
 
 
+def _ari_should_renew(
+    config: configuration.NamespaceConfig, lineage: interfaces.RenewableCert) -> bool:
+    """Return true if ARI says the certificate should be renewed.
+
+    Queries the ACME server for the certificate's Renewal Info, picks a random
+    time within the Suggested Window, and returns True if the selected time is
+    in the past.
+    """
+    # We need to use OCSP, so if we failed to import it, just bail out now.
+    if ocsp is None:
+        logger.info("Skipping ARI check because can't compute ARI path")
+        return False
+
+    # Compute an OCSP request using the given certificate and its issuer.
+    with open(lineage.cert_path, 'rb') as cert_file:
+        cert = x509.load_pem_x509_certificate(cert_file.read(), default_backend())
+    with open(lineage.chain_path, 'rb') as chain_file:
+        issuer = x509.load_pem_x509_certificate(chain_file.read(), default_backend())
+    builder = ocsp.OCSPRequestBuilder()
+    builder = builder.add_certificate(cert, issuer, hashes.SHA1())
+    ocspRequest = builder.build()
+
+    # Construct the ARI path from the OCSP CertID sequence.
+    key_hash = ocspRequest.issuer_key_hash.hex()
+    name_hash = ocspRequest.issuer_name_hash.hex()
+    serial = hex(ocspRequest.serial_number)[2:]
+    path = f"{key_hash}/{name_hash}/{serial}"
+
+    # Construct an ACME client, violating like three layers of abstraction to do so.
```

Comment:
As it seems you are aware, this is not my favorite solution ever written. Was it intended to be temporary? Can `le_client` not just be passed in from above? We've definitely already called `_init_le_client` before now, right?

Reply:
Yeah it's gross and I'd love to replace it.

As far as I can tell, there are two codepaths that end up calling `renewal.should_renew()`. The first *is* called after an `_init_le_client` call inside `main.run()` or `main.certonly()`, but would require passing that client from there all the way through
- `_find_cert()`
- `_find_lineage_for_domains_and_certname`
- `_find_lineage_for_domains`
- `_handle_identical_cert_request`
- finally to `renewal.should_renew`.

The second codepath is `main.renew()` --> `renewal.handle_renewal_request()` --> `renewal.should_renew()`, and that code path does *not* call `_init_le_client` prior to reaching this code, and in fact does the same unfortunate "import `main` inside a function so it isn't technically an import cycle" dance there:
https://github.com/certbot/certbot/blob/d2578e05e7cfd6e408dc45b6e48d0c42685c014d/certbot/certbot/_internal/renewal.py#L458-L475

Due to the first codepath requiring additional arguments to a lot of functions, and the second codepath not having access to a client at all, I went with this for now. I'd really love to not land this version though 🤣

Reply:
Ok, thinking out loud here. I see why it went directly in `should_renew`, but I think logically it might belong more clearly in `RenewableCert.should_autorenew` ([link](https://github.com/certbot/certbot/blob/d20e42562cdb9cfd46fa5cd592a489bc16a5de4c/certbot/certbot/_internal/storage.py#L942)). That's where we also do the ocsp check, and have our current internal logic for if it should be renewed, which ideally this proposal would replace. The reason we can stick ocsp in there without going through any sort of acme client is because we grab the ocsp url directly from the certificate. Theoretically there's no actual reason the static renewal url *needs* to be grabbed from the directory each time live. So if we've already reconstructed the `RenewableCert` object, which we definitely have because it's passed into `should_renew`, we could....cache it in the renewal info (`/etc/letsencrypt/renewal/example.com.conf`). Which I also don't love! 

Oh! ~~We could also just instantiate a `certbot._internal.client.Client` [directly](https://github.com/certbot/certbot/blob/d2578e05e7cfd6e408dc45b6e48d0c42685c014d/certbot/certbot/_internal/client.py#L245)!~~ We don't *actually* need an authenticator *or* installer for this particular purpose.

Actually we may not want to even bother with a `certbot...client`. Like, it's a get, we're not signing POSTs, we could just construct an acme client network and a clientv2. In fact this could all be a nice acme util method, I give you a server directory url (which we DO have at all these points), you give back the renewal info url, much like grabbing it out of the cert for ocsp.

Really the key here is, since this is in the directory and as I understand it not actually authenticated, how much do we care about loading/creating the account information? Because if we *don't*, that opens things up to be directly instantiated to a greater extent. The logic in main is mostly about making sure we're all nicely set up with accounts to be ready to do whatever acme stuff we might need to do. But we don't *actually* need to do that here.

Reply:
Basically, I see two main pathways for doing it unhackily:

- Given that if the answer is "yes" we'll eventually want a whole client properly instantiated, do some refactoring, such as changing where checks happen or how initialized clients are made available (could be moved out of main, which certainly could use literally any common code removed from it).
- Think of this like OCSP and consider it an unauthenticated off-path side request where we just happen to be grabbing the URL from the acme server, and stick in a nice little bypass to get that information more directly.

Open to others, of course.

Reply:
I've gone the route of adding an `AriChecker` which parallels the existing OCSP `RevocationChecker` and which can operate without needing a full ACME account or client. As part of doing so, I've moved the check out of `renewal.should_renew()` and into `RenewableCert.should_autorenew()` as you suggest. I definitely like it better this way.

Only issue now is that the new `ari.AriChecker` needs the base URL, which means we need to query the directory for that, which means we need the directory URL... which I haven't figured out how to access from within `RenewableCert.should_autorenew()` yet. But I'm sure I'll figure it out soon enough! If nothing else, I can try construct a client from the config, skipping the authenticator and installer, and use than unauthenticated client to request the directory.

Reply:
Yeah, I seem to remember hitting that wall as well when I last looked, I don't think there is currently a clean way to do that. If that doesn't end up being too gross, what you suggested sounds reasonable. I briefly considered that we could cache either the whole directory or just the relevant URL (in the renewal config or elsewhere) when we first get the cert, but I don't love that. Even if we need to throw a utility method in `acme` to just grab the contents of the directory, that's still not too bad in terms of building up specific cruft.